pkg/controllers/testing/table.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2019 the original author or authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"strings"
+	gotesting "testing"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/projectriff/system/pkg/tracker"
+)
+
+// Testcase holds a single row of a table test.
+type Testcase struct {
+	// Name is a descriptive name for this test suitable as a first argument to t.Run()
+	Name string
+
+	// Focus is true if and only if only this and any other focussed tests are to be executed.
+	// If one or more tests are focussed, the overall table test will fail.
+	Focus bool
+
+	// Skip is true if and only if this test should be skipped.
+	Skip bool
+
+	// Objects holds the kubernetes objects which are present at the onset of reconciliation
+	Objects []runtime.Object
+
+	// Key identifies the object to be reconciled
+	Key types.NamespacedName
+
+	// ExpectTrack holds the ordered list of Track calls expected during reconciliation
+	ExpectTrack []TrackRequest
+
+	// ExpectCreate holds the ordered list of objects expected to be created during reconciliation
+	ExpectCreate []runtime.Object
+
+	// ExpectStatusUpdated holds the ordered list of objects whose status is updated during reconciliation.
+	ExpectStatusUpdated []runtime.Object
+
+	// ExpectErr is true if and only if reconciliation is expected to return an error
+	ExpectErr bool
+
+	// ExpectErrMessage is the error message expected if ExpectErr is true
+	ExpectErrMessage string
```

Comment:
Not sure how this would really improve things. Unlike in the cli test which are looking for "silent" errors, the only thing we have to go on is the error message, so if we had a verify method, it would end up with similar strings to the current approach.

Reply:
Agree we should verify the rich error value.

Reply:
Added To Do item to https://github.com/projectriff/system/issues/212.